packages/slice-machine/src/features/sync/AutoSyncProvider.tsx

```diff
@@ -0,0 +1,247 @@
+  createContext,
+  useCallback,
+  useContext,
+  useEffect,
+  useMemo,
+} from "react";
+import {
+  useOnChange,
+  useStableCallback,
+} from "@prismicio/editor-support/React";
+
+import { AuthStatus } from "@src/modules/userContext/types";
+import useSliceMachineActions from "@src/modules/useSliceMachineActions";
+import { getState } from "@src/apiClient";
+import {
+  Environment,
+  isUnauthenticatedError,
+} from "@slicemachine/manager/client";
+import { useNetwork } from "@src/hooks/useNetwork";
+import { useAuthStatus } from "@src/hooks/useAuthStatus";
+import {
+  ChangedCustomType,
+  ChangedSlice,
+} from "@lib/models/common/ModelStatus";
+
+import { useActiveEnvironment } from "../environments/useActiveEnvironment";
+import { ActionQueueStatus, useActionQueue } from "../../hooks/useActionQueue";
+import { pushChanges } from "./actions/pushChanges";
+import { useUnSyncChanges } from "./useUnSyncChanges";
+import { fetchUnSyncChanges } from "./fetchUnSyncChanges";
+
+export type AutoSyncStatus =
+  | "not-active"
+  | "offline"
+  | "not-logged-in"
+  | "syncing"
+  | "synced"
+  | "failed";
+
+type AutoSyncContext = {
+  autoSyncStatus: AutoSyncStatus;
+  syncChanges: (args?: SyncChangesArgs) => void;
+};
+
+type SyncChangesArgs = {
+  environment?: Environment;
+  loggedIn?: boolean;
+  changedCustomTypes?: ChangedCustomType[];
+  changedSlices?: ChangedSlice[];
+};
+
+const AutoSyncContextValue = createContext<AutoSyncContext | undefined>(
+  undefined,
+);
+
+export const AutoSyncProvider: FC<PropsWithChildren> = (props) => {
+  const { children } = props;
+  const { unSyncedCustomTypes, unSyncedSlices } = useUnSyncChanges();
+  const isOnline = useNetwork();
+  const authStatus = useAuthStatus();
+  const { refreshState, pushChangesSuccess } = useSliceMachineActions();
+  const stableRefreshState = useStableCallback(refreshState);
+  const stablePushChangesSuccess = useStableCallback(pushChangesSuccess);
+  const { activeEnvironment } = useActiveEnvironment();
+  const { setNextAction, actionQueueStatus } = useActionQueue({
+    actionQueueStatusDelay: 0,
+    errorMessage:
+      "Failed to sync changes. Check your browser's console for more information.",
+  });
+
+  const syncChanges = useCallback(
+    (args: SyncChangesArgs = {}) => {
+      const {
+        // We default to the active environment if not provided.
+        // This is useful when we want to sync changes right after an environment switch.
+        environment = activeEnvironment,
+
+        // We default to a full user logged in with internet access if not provider.
+        // This is useful when we want to sync changes right after the user logs in.
+        loggedIn = isOnline && authStatus === AuthStatus.AUTHORIZED,
+      } = args;
+
+      console.log("1. Just calling syncChanges");
+
+      if (!loggedIn || environment?.kind !== "dev") {
+        return;
+      }
+
+      setNextAction(async () => {
+        console.log("2. Inside setNextSave");
```

Comment:
🙈

Reply:
Removed, console.log for QA on the alpha